### ISSUE TITLE
Strange CollectionFixture issues since 2.3.0 rc1 (including final)

### ISSUE THREAD

**Issue by jherby2k** [xunit/xunit#1480]
I've had to revert to beta5 as rc1 is throwing null exceptions in my CollectionFixture constructor.

Seems like an issue with the code i'm calling except it only happens with rc1 of xunit, not any prior version.

Try running the tests in https://github.com/jherby2k/AudioWorks. The latest version reverts to beta5 and does not exhibit the issue.

InitialSessionState.CreateDefault() in the ModuleFixture constructor (in the AudioWorks.Commands.Tests project) returns null under rc1, and a valid object under any prior version. InitialSessionState.CreateDefault() is from the latest PowerShell Core beta.

**Comment by bradwilson** [comment]
I am unable to work with this repro because of package restore failures:

```
C:\Dev\repros\xunit-1480\AudioWorks\Extensions\AudioWorks.Extensions.Id3\AudioWorks.Extensions.Id3.csproj : error NU1101: Unable to find package csid3lib. No packages exist with this id in source(s): CliFallbackFolder, Microsoft Visual Studio Offline Packages, nuget.org [C:\Dev\repros\xunit-1480\AudioWorks\AudioWorks.sln]
C:\Dev\repros\xunit-1480\AudioWorks\AudioWorks.Commands.Tests\AudioWorks.Commands.Tests.csproj : error NU1101: Unable to find package Microsoft.PowerShell.Commands.Diagnostics. No packages exist with this id in source(s): CliFallbackFolder, Microsoft Visual Studio Offline Packages, nuget.org [C:\Dev\repros\xunit-1480\AudioWorks\AudioWorks.sln]
C:\Dev\repros\xunit-1480\AudioWorks\AudioWorks.Commands.Tests\AudioWorks.Commands.Tests.csproj : error NU1101: Unable to find package Microsoft.PowerShell.SDK. No packages exist with this id in source(s): CliFallbackFolder, Microsoft Visual Studio Offline Packages, nuget.org [C:\Dev\repros\xunit-1480\AudioWorks\AudioWorks.sln]
C:\Dev\repros\xunit-1480\AudioWorks\AudioWorks.Commands.Tests\AudioWorks.Commands.Tests.csproj : error NU1101: Unable to find package Microsoft.WSMan.Management. No packages exist with this id in source(s): CliFallbackFolder, Microsoft Visual Studio Offline Packages, nuget.org [C:\Dev\repros\xunit-1480\AudioWorks\AudioWorks.sln]
```

Are you missing a `NuGet.Config` file?

**Comment by jherby2k** [comment]
Yeah sorry, wasn’t ready for public consumption… I just checked in two config files.

Sent from Mail for Windows 10

**Comment by jherby2k** [comment]
Issue seems resolved with rc2.

**Comment by jherby2k** [comment]
Nah, issue seems intermittent somehow even with 2.3.0 final. Never happens with any of the betas, or 2.2.0.

**Comment by jherby2k** [comment]
Please check out latest https://github.com/jherby2k/AudioWorks to see issue, then downgrade to beta-5 to fix it.

**Comment by bradwilson** [comment]
The `csid3lib` NuGet package does not restore. There is no such package on NuGet: https://www.nuget.org/packages/csid3lib

**Comment by jherby2k** [comment]
Grr no it’s on myget. Not sure why visual studio keeps choking on the nuget.config. It is in the solution root. If you put it into nuget options in visual studio it works fine

**Comment by bradwilson** [comment]
Changing NuGet options in Visual Studio changes your global settings. You have to updated NuGet.Config by hand.

**Comment by jherby2k** [comment]
You can also just disable the audioworks.extensions.id3 project and ignore the mp3 test failures

**Comment by jherby2k** [comment]
Gah, you’re right. It was missing… okay its fixed for sure if you update to the latest commit.

Sent from Mail for Windows 10

**Comment by bradwilson** [comment]
Which runner are you using? I've been testing with `dotnet xunit`.

I upgraded all packages to `2.3.1-rc2-build3844`. The API tests are all passing.

The commands tests have some failures, but they are not null reference exceptions. They look like this:

![Commands test failure](https://user-images.githubusercontent.com/16944/31810857-25849bee-b532-11e7-9d20-d4c51a636a29.png)

I'm not sure whether this is your original issue or not.

**Comment by jherby2k** [comment]
It is the command tests failing but that’s not the same issue. I am using the vs runner. Let me try again with the latest version and dotnet xunit

**Comment by jherby2k** [comment]
Well that’s interesting… I have never tried dotnet xunit. It fails for me too regardless of xunit version. Might be related, might not be.

Anyway, with the visual studio runner, there should be 0 failures. Except that upgrading to 2.3.1 rc2 still causes all the command tests to fail as discussed.

Weirdly, the tests pass the first time after upgrading xunit, and then fail on subsequent runs until I downgrade to beta 5.

**Comment by jherby2k** [comment]
Also “dotnet test” passes fine. Not sure what’s the problem with “dotnet xunit”… 

Sent from Mail for Windows 10

From: Jeremy Herbison
Sent: Friday, October 20, 2017 9:09 AM
To: xunit/xunit; xunit/xunit
Cc: State change
Subject: RE: [xunit/xunit] Strange CollectionFixture issues since 2.3.0rc1(including final) (#1480)

Well that’s interesting… I have never tried dotnet xunit. It fails for me too regardless of xunit version. Might be related, might not be.

Anyway, with the visual studio runner, there should be 0 failures. Except that upgrading to 2.3.1 rc2 still causes all the command tests to fail as discussed.

Weirdly, the tests pass the first time after upgrading xunit, and then fail on subsequent runs until I downgrade to beta 5.

**Comment by jherby2k** [comment]
This is still an issue with xunit 2.3.1. The commands tests fail the _second_ time you run the tests, after upgrading xunit. This is with the visual studio runner.

**Comment by jherby2k** [comment]
Since “dotnet test” works fine but “dotnet xunit” fails with “command not found” exceptions I thought maybe it was related to #1509, but upgrading to xunit 2.4 beta 1 doesn’t help.

Sent from Mail for Windows 10

From: Jeremy Herbison
Sent: Friday, October 27, 2017 3:17 PM
To: xunit/xunit
Cc: Brad Wilson
Subject: RE: [xunit/xunit] Strange CollectionFixture issues since 2.3.0rc1(including final) (#1480)

This is still an issue with xunit 2.3.1. The commands tests fail the _second_ time you run the tests, after upgrading xunit. This is with the visual studio runner.

**Comment by bradwilson** [comment]
> The tests pass the first time after upgrading xunit, and then fail on subsequent runs until I downgrade to beta 5.

That sounds like you may be running afoul of the fact that Visual Studio is leaving your assembly loaded and reusing it if you re-run tests. Does it continue to exhibit the failure if you do a Clean and then Build, and _then_ re-run the tests?

**Comment by jherby2k** [comment]
Well, after the first post-upgrade test run (which works) I immediately run it again and the command tests all fail. If I then reboot, clean and rebuild, then run it again… it still fails. And continues to until I downgrade to beta 5. Beta 5 is rock solid every time.

I have confirmed that “Dotnet test” in the audioworks.commands.tests project behaves exactly the same. You should be able to reproduce this.

**Comment by jherby2k** [comment]
Blargh, there was a DLL version conflict in some of the PowerShell Core nuget packages. I changed the way I deploy the module from a "dotnet publish” to a simple reference and now the issue is gone.

Not sure why the xunit upgrade caused that to be noticeable, but whatever. Sorry to take your time! Thanks for all your hard work!

Jeremy
